gnucash/import-export/qif-imp/qif-file.scm

```diff
@@ -1083,3 +1133,40 @@
               (test-results (cdr results))))))
 
   (if results (test-results results) #f))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;  qif-file:parse-price-line
+;;
+;;  Takes an expected price line and returns a qif-price object
+;;
+;;  We are parsing "SYMBOL",value,"DATE"
+;;    where symbol is the alpha name of the security symbol,
+;;    value can be a float like 42.50 or a fractional expression like 42 1/2
+;;    and date can be the typical Quicken date formats
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(define (qif-file:parse-price-line line)
+  (let ((current-xtn (make-qif-price)))
+    (if (not (false-if-exception
+        (let* ((symbol (string-trim-both (string-trim-both (car (string-split line #\,)) char-set:punctuation)))
+               (price (cadr (string-split line #\,)))
+               (fracprice (string-split price #\space))
+               (date (string-trim-both (string-trim-both (caddr (string-split line #\,)) char-set:punctuation))))
+
+          (qif-price:set-symbol! current-xtn symbol)
+          (qif-price:set-date! current-xtn date)
+
+          ;; If we have a fractional portion, convert it to float
+          (if (> (length fracprice) 1)
+              (let ((total (+ (string->number (car fracprice))
+                 (string->number (cadr fracprice)))))
+              (set! price (format #f "~f" total))))
+
+          (qif-price:set-share-price! current-xtn price)
+
+          ;; A blank entry will not necessarily throw an exception, but is invalid
+          (if (or (string-null? symbol) (or (string-null? price) (string-null? date)))
+            (set! current-xtn #f)))))
+       (set! current-xtn #f))
+    current-xtn))
```

Comment:
I'll rewrite it to avoid exceptions and use pattern matching instead. It also includes a home grown string to numeric parser which is too lenient. "1 3/4" is converted to 1.75 and "1 2" will be converted to "3" without erroring out. Does @jralls know of such a parser that can handle "1" "2 3/4" well and decline to parse "4 5" within the engine?

Reply:
It's not just too lenient, it generates floating point when it should create either gnc_numeric or Scheme exact rational. 

Try `string-to-gnc-numeric` for parsing, that's what pretty much everything else uses. I don't know what it does with internal whitespace when the second number isn't a fraction.

Reply:
https://github.com/Gnucash/gnucash/pull/1756 if gnc_numeric_to_string is upgraded.

Reply:
Hmm.   That does not drop in and just work.; the import fails. The resulting strings from this function are, I presume, scheme rational numbers (i.e. I see "21/20" and so on).  But qif-file:parse-fields wants to reprocess all transactions, expecting string input.   And that seems to boil down to a call to  qif-parse:parse-number/format, which, afaict, does not handle this format - it has decimal, comma, and integer types.

Reply:
How annoying. The qif-parser was written 20 years ago.

```scheme
modified   gnucash/import-export/qif-imp/qif-file.scm
@@ -832,7 +832,7 @@
       (start-sub 1)
       (check-and-parse-field
        qif-price:share-price qif-price:set-share-price! gnc-numeric-equal
-       qif-parse:check-number-format '(decimal comma)
+       qif-parse:check-number-format '(decimal comma rational)
        qif-parse:parse-number/format (qif-file:prices self)
        qif-parse:print-number
        'guess-on-ambiguity add-error 'share-price
@@ -883,7 +883,7 @@
       (start-sub 0.1)
       (check-and-parse-field
        qif-xtn:share-price qif-xtn:set-share-price! gnc-numeric-equal
-       qif-parse:check-number-format '(decimal comma)
+       qif-parse:check-number-format '(decimal comma rational)
        qif-parse:parse-number/format (qif-file:xtns self)
        qif-parse:print-number
        'guess-on-ambiguity add-error 'share-price
modified   gnucash/import-export/qif-imp/qif-parse.scm
@@ -404,6 +404,11 @@
   (and regexp-enabled?
        (make-regexp "^ *[$]?[+-]?[$]?[0-9]+[+-]?$|^ *[$]?[+-]?[$]?[0-9]?[0-9]?[0-9]?([\\.'][0-9][0-9][0-9])*(,[0-9]*)?[+-]? *$|^ *[$]?[+-]?[$]?[0-9]+,[0-9]*[+-]? *$")))
 
+;; eg 1/2 or 32/15 or 4
+(define rational-regexp
+  (and regexp-enabled?
+       (make-regexp "^-?[0-9]+(/[0-9]+|)$")))
+
 ;; eg 456 or 123
 (define integer-regexp
   (and regexp-enabled?
@@ -419,6 +424,7 @@
   (define numtypes-alist
     (list (cons 'decimal decimal-radix-regexp)
           (cons 'comma comma-radix-regexp)
+          (cons 'rational rational-regexp)
           (cons 'integer integer-regexp)))
   (filter (lambda (fmt) (regexp-exec (assq-ref numtypes-alist fmt) value-string))
           possible-formats))
@@ -439,7 +445,7 @@
                         ((comma) (gnc:string-replace-char
                                   (gnc:string-delete-chars filtered-string ".")
                                   #\, #\.))
-                        ((integer) filtered-string)))
+                        ((integer rational) filtered-string)))
          (num (or (string->number (string-append "#e" read-string)) 0)))
     (if has-minus? (- num) num)))
```

Reply:
Good to know that I've learned everything I know about Guile from that code base.  If only this was 2001, I'd be so cool  :-).

Reply:
I've in theory pushed your code into this PR, and, if I rebase to include the new code, it seems to work for me.